### DISCUSSION TITLE
Importing spacy is slow

### DISCUSSION THREAD

**Discussion by gaborbernat** [explosion/spaCy#11158] [Help: Coding & Implementations]
Just importing spacy 3.4.0 into Python takes 1.4 seconds. Is this expected? Can we improve on this?

```
❯ hyperfine ".tox/dev/bin/python -c 'import spacy'"
Benchmark 1: .tox/dev/bin/python -c 'import spacy'
  Time (mean ± σ):      1.471 s ±  0.106 s    [User: 1.762 s, System: 0.351 s]
  Range (min … max):    1.281 s …  1.645 s    10 runs
```

I'm trying to do simple tokenization:

```python
from spacy.lang.en import English
from spacy.tokenizer import Tokenizer

nlp = English()
tokenizer = Tokenizer(nlp.vocab)
res = list(tokenizer(text))
```

But just importing spacy takes 1.4 seconds that's really noticeable for a CLI application, everything else is <100ms, so after import is fast again.

Here's the import waterfall dump, https://gist.github.com/gaborbernat/682bf98bd9b111d40e86ec7f92625844, which you can view by copying to http://www.softwareishard.com/har/viewer

**Comment by polm** [comment]
This is definitely something we could improve on, though startup time in general is something Python isn't great at. We've put most of our effort into making sure processing is fast for documents after process startup. In the typical case where you're using a model and not just the tokenizer, startup is going to take some time in any case. For a CLI app, if you need spaCy I would recommend running it in a background process. 

Also note that startup time can be affected by what entry points are available in your current Python environment. A stripped-down venv should be faster than an environment with many packages installed. You can profile Python imports and startup time using tools [here](https://pythondev.readthedocs.io/startup_time.html).

Also I tried the preview site you mentioned but the units appear to be off - maybe I'm reading it wrong but it seems to say startup takes 20m?

**Reply by gaborbernat** [reply]
I did run my tests in a fresh environment with just spacy installed into it, so extra entry points did make an impact. I've used https://github.com/asottile/importtime-waterfall to profile the import times, which generates the HAR rather than the PNG in the article you've linked. Yes, the numbers are off the unit in the HAR, but otherwise accurate relative to each other component.
